### PR TITLE
scheduler: DeviceShare supports allocating by hints

### PR DESCRIPTION
**Pull request by eahydra**
### Ⅰ. Describe what this PR does

Enhance the capabilities of the DeviceShare plugin, support joint-allocation, and support allocation by hint api(more detail please refer to [20230803-device-allocate-hint-apis.md](https://github.com/koordinator-sh/koordinator/blob/main/docs/proposals/scheduling/20230803-device-allocate-hint-apis.md)).

This PR involves some changes in code structure, which affects some UTs, so it looks relatively large, but the core logic is all in device_allocator.go.

### V. Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `make test`

### CODE REVIEW

File: pkg/scheduler/plugins/deviceshare/device_allocator.go [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+}
+
+type AutopilotAllocator struct {
+	state                     *preFilterState
+	nodeDevice                *nodeDevice
+	node                      *corev1.Node
+	pod                       *corev1.Pod
+	scorer                    *resourceAllocationScorer
+	numaNodes                 bitmask.BitMask
+	requestsPerInstance       map[schedulingv1alpha1.DeviceType]corev1.ResourceList
```
Comment (hormes) [inline pkg/scheduler/plugins/deviceshare/device_allocator.go:66]: 这里为啥叫 requestsPerInstance，Instance 指代的是什么？

File: pkg/scheduler/plugins/deviceshare/device_allocator.go [hunk 2]
```diff
@@ -0,0 +1,504 @@
...
+	deviceAllocations, status := a.allocateByTopology(requestCtx, nodeDevice, topologyGuide, primaryDeviceType, jointAllocate, jointAllocate.DeviceTypes[1:])
+	if jointAllocate.RequiredScope == apiext.SamePCIeDeviceJointAllocateScope {
+		if status.IsSuccess() {
+			status = a.validateJointAllocation(jointAllocate, nodeDevice, deviceAllocations)
+			if !status.IsSuccess() {
+				return nil, status
+			}
+		}
+	} else if !status.IsSuccess() {
+		status = nil
```
Comment (hormes) [inline pkg/scheduler/plugins/deviceshare/device_allocator.go:209]: 不要强 PCIe 相同时，为啥需要再这里忽略错误？

File: pkg/scheduler/plugins/deviceshare/device_allocator.go (hunk 1 above)
Comment (eahydra) [reply pkg/scheduler/plugins/deviceshare/device_allocator.go:66]: instance是device instance，设备实例。

**Review by jasonliu747 — COMMENTED** [review]
/lgtm

**Review by hormes — APPROVED** [review]

**Pull request merged**